src/read/read_cache.rs

```diff
@@ -0,0 +1,120 @@
+use std::boxed::Box;
+use std::cell::RefCell;
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::convert::TryInto;
+use std::io::{Read, Seek, SeekFrom};
+use std::mem;
+
+use crate::read::ReadRef;
+
+/// An implementation of `ReadRef` for data in a stream that implements
+/// `Read + Seek`.
+///
+/// Contains a cache of read-only blocks of data, allowing references to
+/// them to be returned. Entries in the cache are never removed.
```

Comment:
Okay, I see now.

As I mentioned earlier, emptying the cache could be possible with a `Ref`/`RefMut`-like return type that has a destructor. This is a bit less ergonomic for users tho and also has a higher cognitive overhead. So not a pure win.